src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -59,6 +108,16 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
                 var document = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
                 _currentDocumentId = document?.Id;
 
+                // Listen for whenever this buffer gets associated with a different workspace/document.
+                _workspaceRegistration = Workspace.GetWorkspaceRegistration(_subjectBuffer.AsTextContainer());
+                _workspaceRegistration.WorkspaceChanged += OnWorkspaceChanged;
+                _workspace = document?.Project.Solution.Workspace;
+
+                if (_workspace != null)
+                {
+                    _workspace.DocumentActiveContextChanged += OnDocumentActiveContextChanged;
+                }
+
                 // Kick off a background task to collect the initial set of diagnostics.
```

Comment:
I don't understand why we wouldn't rerun this process any time the document ID that we're watching changes, which would seem like it would address the race concerns being brought up by the background filtering. Put another way, "we were created and started looking at a DocumentID" and "context changed, you're watching a different DocumentID" really should be the same codepaths.

One pattern you might be able to use then is whenever the document ID we're tracking changes, create a new CancellationTokenSource a la _initialDiagnosicsCancellationSource and cancel the old one. That's captured by the task that does the initial population, but also any tasks that did background filtering _and_ the resulting posts to the UI thread. The UI thread checks the local copy of that token, so if the document is changed, any in-flight queued work sees the token is cancelled and knows not to update anything.